### PR TITLE
Add RVM/interpreter incompatibility tests and compiler error detection

### PR DESCRIPTION
**Pull request by anakrish**
- Add 12 new YAML test files covering RVM/interpreter edge cases: aggregates, base/virtual docs, data deref, default keyword, every quantifier, advanced functions, partial doc constants, partial objects, planner IR, ref heads, type builtins, walk builtin

- Add compiler errors for unsupported patterns:
  - FunctionDefaultsUnsupported: default f(x) := value
  - ComprehensionInDefaultUnsupported: default p := [x | ...]
  - MultiValueRefHeadUnsupported: p[q].r or p[q] contains r

- Trim OPA_TODO_FOLDERS from 14 blanket-skipped folders to 1 (withkeyword) by replacing folder-level skips with:
  - compiler_limitation_reason(): auto-skips on 11 known compiler error patterns (with keyword, walk, function defaults, comprehension defaults, multi-value ref heads, undefined variable, unknown function, etc.)
  - KNOWN_RVM_MISMATCH_NOTES: 32 entries for runtime RVM bugs that compile successfully but produce incorrect results (every false-case, base/virtual merge, suffix lookup, dynamic cross-rule bracket lookup, etc.)

- Net effect: 290 more OPA tests now actively validate RVM correctness

### CODE REVIEW

File: src/languages/rego/compiler/rules.rs [hunk 1]
```diff
@@ -340,6 +349,15 @@ impl<'a> Compiler<'a> {
 
                     let (key_expr, value_expr) = match head {
                         RuleHead::Compr { refr, assign, .. } => {
+                            // Detect multi-key ref heads: p[q].r or p[q][r]
+                            if let Expr::RefBrack { refr: inner, .. } = refr.as_ref() {
+                                if Self::ref_contains_bracket(inner.as_ref()) {
+                                    return Err(
+                                        CompilerError::MultiValueRefHeadUnsupported.at(span)
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/languages/rego/compiler/rules.rs]: Extended to cover both p[q].r and p[q][r]

File: tests/rvm/rego/cases/ref_heads.yaml [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    query: data.test.matrix
+    want_result:
+      a:
+        x: 11
+        y: 21
+      b:
+        x: 12
+        y: 22
+    skip: true
+
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/ref_heads.yaml]: Removed skip and changed to want_error.

File: tests/rvm/rego/cases/ref_heads.yaml [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+          u: true
+      q:
+        r:
+          t: true
+          u: true
+        s:
+          t: true
+          u: true
+    skip: true
+
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/ref_heads.yaml]: removed skip and changed to want_error.

File: tests/rvm/rego/cases/ref_heads.yaml [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    query: data.test.allowed
+    want_result:
+      alice:
+        set!:
+          - "cache"
+          - "db"
+      bob:
+        set!:
+          - "db"
+    skip: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/ref_heads.yaml]: removed skip and converted to want_error.

File: tests/rvm/rego/cases/ref_heads.yaml [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    modules:
+      - |
+        package test
+        matrix[row][col] := value if {
+          some row, rv in input.rows
+          some col, cv in input.cols
+          value := rv + cv
+        }
+    query: data.test.matrix
+    want_error: "not a valid rule path"
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/ref_heads.yaml:26]: For these scenarios. The "not a valid rule path" fires before  the specific error is reached. The expectation is correct.

File: tests/rvm/rego/cases/ref_heads.yaml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    modules:
+      - |
+        package test
+        cube[a][b][c] := true if {
+          some a in input.xs
+          some b in input.ys
+          some c in input.zs
+        }
+    query: data.test.cube
+    want_error: "not a valid rule path"
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/ref_heads.yaml:45]: Same as above.

File: tests/rvm/rego/cases/ref_heads.yaml [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    modules:
+      - |
+        package test
+        allowed[user][resource] if {
+          some grant in input.grants
+          user := grant.user
+          resource := grant.resource
+        }
+    query: data.test.allowed
+    want_error: "not a valid rule path"
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/ref_heads.yaml:126]: same as above.

File: tests/rvm/rego/cases/base_and_virtual_docs.yaml [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  # Query a suffix path on a complete rule returning an object
+  - note: base_virtual/suffix_on_complete_rule
+    data: {}
+    modules:
+      - |
+        package test
+        user := {"name": "Alice", "role": "admin", "level": 5}
+    query: data.test.user.role
+    want_result: "admin"
+    skip: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/base_and_virtual_docs.yaml:19]: These will be enabled once the fixes are in.

File: tests/rvm/rego/cases/base_and_virtual_docs.yaml [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    modules:
+      - |
+        package test
+        users[name] := info if {
+          some name
+          info := data.people[name]
+        }
+    query: data.test.users.alice.role
+    want_result: "admin"
+    skip: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/base_and_virtual_docs.yaml:36]: these will be enabled once fixes are in.

File: tests/rvm/rego/cases/base_and_virtual_docs.yaml [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      test:
+        config:
+          base_key: "from_data"
+    modules:
+      - |
+        package test.config
+        virtual_key := "from_rule"
+    query: data.test.config.virtual_key
+    want_result: "from_rule"
+    skip: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/base_and_virtual_docs.yaml:52]: These will be enabled once fixes are in. The RVM only skip is not worth the effort for now.

File: tests/rvm/rego/cases/base_and_virtual_docs.yaml [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+          "database": {
+            "connection": {
+              "host": "localhost",
+              "port": 5432
+            }
+          }
+        }
+    query: data.test.settings.database.connection.host
+    want_result: "localhost"
+    skip: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/base_and_virtual_docs.yaml:70]: same as above.

File: tests/rvm/rego/cases/base_and_virtual_docs.yaml [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    modules:
+      - |
+        package test
+        employees := [
+          {"id": 1, "name": "Alice"},
+          {"id": 2, "name": "Bob"}
+        ]
+    query: data.test.employees[0].name
+    want_result: "Alice"
+    skip: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/base_and_virtual_docs.yaml:84]: same as above.

File: tests/rvm/rego/cases/default_keyword.yaml [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+      - |
+        package test
+        default config := data.fallback_config
+        config := input.override if {
+          input.override
+        }
+    query: data.test.config
+    want_result:
+      timeout: 60
+    skip: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/default_keyword.yaml:27]: same as above.

File: tests/rvm/rego/cases/default_keyword.yaml [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        default users["guest"] := {"role": "viewer"}
+        users[name] := info if {
+          some user in input.user_list
+          name := user.name
+          info := {"role": user.role}
+        }
+    query: data.test.users.guest
+    want_result:
+      role: "viewer"
+    skip: true
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/default_keyword.yaml:65]: they will be enabled after code fix.

File: tests/rvm/rego/cases/walk_builtin.yaml [hunk 15]
```diff
@@ -0,0 +1,87 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+# Walk Builtin Test Suite
+# Tests the walk() builtin for recursive tree traversal.
+# Tracked under the 'walkbuiltin' OPA_TODO_FOLDER.
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/walk_builtin.yaml]: Fixed.

File: src/languages/rego/compiler/program.rs [hunk 16]
```diff
@@ -134,6 +135,31 @@ impl<'a> Compiler<'a> {
             rule_infos_map.insert(rule_index as usize, rule_info);
         }
 
+        // Validate unsupported default rule patterns before evaluation.
+        for (rule_path, default_infos) in self.policy.inner.default_rules.iter() {
+            if !self.rule_index_map.contains_key(rule_path) {
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/languages/rego/compiler/program.rs:145]: Added comment to explain the intentional guard.

File: tests/rvm/rego/cases/aggregates.yaml [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        banana: {price: 1}
+        cherry: {price: 5}
+    modules:
+      - |
+        package test
+        prices[name] := price if {
+          some name
+          price := data.items[name].price
+        }
+        most_expensive := max([v | some v in prices])
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/aggregates.yaml]: fixed.

File: tests/rvm/rego/cases/aggregates.yaml [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+        sales: {headcount: 5}
+        ops: {headcount: 8}
+    modules:
+      - |
+        package test
+        headcounts[dept] := hc if {
+          some dept
+          hc := data.departments[dept].headcount
+        }
+        total_headcount := sum([hc | some hc in headcounts])
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/aggregates.yaml]: fixed.

File: tests/opa.rs [hunk 19]
```diff
@@ -262,24 +249,175 @@ fn eval_rule_with_rvm(case: &TestCase, is_rego_v0_test: bool, rule_path: &str) -
...
+            "multi-value ref heads unsupported",
+        ),
+    ];
+    for (pattern, reason) in patterns {
+        if err_str.contains(pattern) {
+            return Some(reason);
+        }
+    }
+    None
 }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/opa.rs:290]: Ok for now.

File: src/languages/rego/compiler/program.rs [hunk 20]
```diff
@@ -134,6 +135,34 @@ impl<'a> Compiler<'a> {
...
+                        Expr::ArrayCompr { .. }
+                        | Expr::SetCompr { .. }
+                        | Expr::ObjectCompr { .. } => {
+                            return Err(CompilerError::ComprehensionInDefaultUnsupported.at(span));
+                        }
+                        _ => {}
+                    }
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply src/languages/rego/compiler/program.rs:164]: already covered in default_keyword.yaml.

File: tests/opa.rs [hunk 21]
```diff
@@ -262,24 +249,175 @@ fn eval_rule_with_rvm(case: &TestCase, is_rego_v0_test: bool, rule_path: &str) -
...
+        (
+            "multi-value ref head rules are not yet supported",
+            "multi-value ref heads unsupported",
+        ),
+    ];
+    for (pattern, reason) in patterns {
+        if err_str.contains(pattern) {
+            return Some(reason);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/opa.rs:288]: not an issue. compiles fine and all tests pass.

File: tests/opa.rs [hunk 22]
```diff
@@ -262,24 +249,175 @@ fn eval_rule_with_rvm(case: &TestCase, is_rego_v0_test: bool, rule_path: &str) -
...
+        "refheads/single-value example, false",
+        "dynamic cross-rule bracket lookup",
+    ),
+];
+
+fn known_rvm_mismatch_reason(note: &str) -> Option<&'static str> {
+    KNOWN_RVM_MISMATCH_NOTES
+        .iter()
+        .find(|(n, _)| *n == note)
+        .map(|(_, reason)| *reason)
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/opa.rs:413]: Compiles and runs fine.

File: tests/rvm/rego/cases/function_advanced.yaml [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+    data: {}
+    modules:
+      - |
+        package test
+        fib(0) := 0
+        fib(1) := 1
+        fib(n) := fib(n - 1) + fib(n - 2) if { n > 1 }
+        result := fib(6)
+    query: data.test.result
+    want_result: 8
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/function_advanced.yaml:39]: Disabled the test as suggested. Both RVM and compiler must be fixed to disallow recursion.

File: tests/rvm/rego/cases/function_advanced.yaml [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+  - note: functions/recursive_factorial
+    data: {}
+    modules:
+      - |
+        package test
+        factorial(0) := 1
+        factorial(n) := n * factorial(n - 1) if { n > 0 }
+        result := factorial(5)
+    query: data.test.result
+    want_result: 120
```
[the remark this answers is not in the document]
Comment (anakrish) [reply tests/rvm/rego/cases/function_advanced.yaml:24]: Disabled the test as suggested. Both RVM and compiler must be fixed to disallow recursion.
